# Name: Nagham Tharwat Ramadan
# ID : 21510350
# TMA TM271

# downloading data

In [1]:
 # Downloading required packages
!pip install ucimlrepo
from ucimlrepo import fetch_ucirepo

# import Required libraries 
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report, roc_auc_score, roc_curve

## Part A  Loading dataset

In [2]:
# Load dataset
heart_disease = fetch_ucirepo(id=45)
heart_data_df = pd.DataFrame(heart_disease.data.features, columns=heart_disease.data.feature_names)
heart_data_df['target'] = heart_disease.data.targets
heart_data_df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   1       145   233    1        2      150      0      2.3   
1     67    1   4       160   286    0        2      108      1      1.5   
2     67    1   4       120   229    0        2      129      1      2.6   
3     37    1   3       130   250    0        0      187      0      3.5   
4     41    0   2       130   204    0        2      172      0      1.4   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   45    1   1       110   264    0        0      132      0      1.2   
299   68    1   4       144   193    1        0      141      0      3.4   
300   57    1   4       130   131    0        0      115      1      1.2   
301   57    0   2       130   236    0        2      174      0      0.0   
302   38    1   3       138   175    0        0      173      0      0.0   

     slope   ca  thal  target  
0        3  0.0   6.0       0  
1        2  3.0   3.0       2  
2        2  2.0   7.0       1  
3        3  0.0   3.0       0  
4        1  0.0   3.0       0  
..     ...  ...   ...     ...  
298      2  0.0   7.0       1  
299      2  2.0   7.0       2  
300      2  1.0   7.0       3  
301      2  1.0   3.0       1  
302      1  NaN   3.0       0  

[303 rows x 14 columns]

## Question No 1: After running the dataset downloading code, you can show the name of the columns, number of rows, number of columns, display first 10 rows and last 10 rows.     

In [3]:
# Show column names
print("Column Names:", heart_data_df.columns)

Column Names: Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')


In [4]:
# Show number of rows and columns
print("Number of Rows:", len(heart_data_df))
print("Number of Columns:", len(heart_data_df.columns))

Number of Rows: 303
Number of Columns: 14


In [5]:
# display first 10 rows
heart_data_df.head(10)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   1       145   233    1        2      150      0      2.3      3   
1   67    1   4       160   286    0        2      108      1      1.5      2   
2   67    1   4       120   229    0        2      129      1      2.6      2   
3   37    1   3       130   250    0        0      187      0      3.5      3   
4   41    0   2       130   204    0        2      172      0      1.4      1   
5   56    1   2       120   236    0        0      178      0      0.8      1   
6   62    0   4       140   268    0        2      160      0      3.6      3   
7   57    0   4       120   354    0        0      163      1      0.6      1   
8   63    1   4       130   254    0        2      147      0      1.4      2   
9   53    1   4       140   203    1        2      155      1      3.1      3   

    ca  thal  target  
0  0.0   6.0       0  
1  3.0   3.0       2  
2  2.0   7.0       1  
3  0.0   3.0       0  
4  0.0   3.0       0  
5  0.0   3.0       0  
6  2.0   3.0       3  
7  0.0   3.0       0  
8  1.0   7.0       2  
9  0.0   7.0       1

In [6]:
# display last 10 rows
heart_data_df.tail(10)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
293   63    1   4       140   187    0        2      144      1      4.0   
294   63    0   4       124   197    0        0      136      1      0.0   
295   41    1   2       120   157    0        0      182      0      0.0   
296   59    1   4       164   176    1        2       90      0      1.0   
297   57    0   4       140   241    0        0      123      1      0.2   
298   45    1   1       110   264    0        0      132      0      1.2   
299   68    1   4       144   193    1        0      141      0      3.4   
300   57    1   4       130   131    0        0      115      1      1.2   
301   57    0   2       130   236    0        2      174      0      0.0   
302   38    1   3       138   175    0        0      173      0      0.0   

     slope   ca  thal  target  
293      1  2.0   7.0       2  
294      2  0.0   3.0       1  
295      1  0.0   3.0       0  
296      2  2.0   6.0       3  
297      2  0.0   7.0       1  
298      2  0.0   7.0       1  
299      2  2.0   7.0       2  
300      2  1.0   7.0       3  
301      2  1.0   3.0       1  
302      1  NaN   3.0       0

## Question No 2: Display basic statistics of the dataset, such as the count, mean, standard deviation etc. Find any missing values 

In [7]:
# show information of columns in dataset
heart_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        299 non-null    float64
 12  thal      301 non-null    float64
 13  target    303 non-null    int64  
dtypes: float64(3), int64(11)
memory usage: 33.3 KB


In [8]:
heart_data_df.shape

(303, 14)

In [9]:
# show all basic statistics of the dataset 
heart_data_df.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.438944    0.679868    3.158416  131.689769  246.693069    0.148515   
std      9.038662    0.467299    0.960126   17.599748   51.776918    0.356198   
min     29.000000    0.000000    1.000000   94.000000  126.000000    0.000000   
25%     48.000000    0.000000    3.000000  120.000000  211.000000    0.000000   
50%     56.000000    1.000000    3.000000  130.000000  241.000000    0.000000   
75%     61.000000    1.000000    4.000000  140.000000  275.000000    0.000000   
max     77.000000    1.000000    4.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  299.000000   
mean     0.990099  149.607261    0.326733    1.039604    1.600660    0.672241   
std      0.994971   22.875003    0.469794    1.161075    0.616226    0.937438   
min      0.000000   71.000000    0.000000    0.000000    1.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    2.000000    0.000000   
75%      2.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    3.000000    3.000000   

             thal      target  
count  301.000000  303.000000  
mean     4.734219    0.937294  
std      1.939706    1.228536  
min      3.000000    0.000000  
25%      3.000000    0.000000  
50%      3.000000    0.000000  
75%      7.000000    2.000000  
max      7.000000    4.000000

# PART B : Data preprocessing 

## Question No 3: Preprocess the data by handling missing values and encoding categorical features (if any).

In [10]:
# Find any missing values
heart_data_df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
target      0
dtype: int64

In [11]:
# Identify columns with missing values
columns_with_missing = heart_data_df.columns[heart_data_df.isnull().any()].tolist()
columns_with_missing

['ca', 'thal']

In [12]:
# Handling missing values
imputer_missvalue= SimpleImputer(strategy='mean') # Replace 'mean' with other strategies if needed
heart_data_df[columns_with_missing] = imputer_missvalue.fit_transform(heart_data_df[columns_with_missing])
heart_data_df.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [13]:
# Convert target column to binary format
heart_data_df['target'] = (heart_data_df['target'] > 0).astype(int)

In [14]:
# Verify the result
print(heart_data_df['target'].value_counts())

0    164
1    139
Name: target, dtype: int64


In [15]:
heart_data_df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   1       145   233    1        2      150      0      2.3      3   
1   67    1   4       160   286    0        2      108      1      1.5      2   
2   67    1   4       120   229    0        2      129      1      2.6      2   
3   37    1   3       130   250    0        0      187      0      3.5      3   
4   41    0   2       130   204    0        2      172      0      1.4      1   

    ca  thal  target  
0  0.0   6.0       0  
1  3.0   3.0       1  
2  2.0   7.0       1  
3  0.0   3.0       0  
4  0.0   3.0       0

In [16]:
# Selecting only the categorical columns
categorical_cols = ['sex', 'cp', 'fbs', 'restecg', 'slope', 'ca', 'thal']
df_categorical = heart_data_df[categorical_cols]
df_categorical

sex  cp  fbs  restecg  slope        ca  thal
0      1   1    1        2      3  0.000000   6.0
1      1   4    0        2      2  3.000000   3.0
2      1   4    0        2      2  2.000000   7.0
3      1   3    0        0      3  0.000000   3.0
4      0   2    0        2      1  0.000000   3.0
..   ...  ..  ...      ...    ...       ...   ...
298    1   1    0        0      2  0.000000   7.0
299    1   4    1        0      2  2.000000   7.0
300    1   4    0        0      2  1.000000   7.0
301    0   2    0        2      2  1.000000   3.0
302    1   3    0        0      1  0.672241   3.0

[303 rows x 7 columns]

In [17]:
# Perform one-hot encoding
encoder = OneHotEncoder()
encoded_data = encoder.fit_transform(df_categorical)

In [18]:
# Convert the encoded data into a DataFrame
encoded_heart_data_df = pd.DataFrame(encoded_data.toarray(), columns=encoder.get_feature_names_out(categorical_cols))

In [19]:
# Concatenate the encoded DataFrame with the original DataFrame, dropping the original categorical columns
df_encoded = pd.concat([heart_data_df.drop(columns=categorical_cols), encoded_heart_data_df], axis=1)
df_encoded
# Now df_encoded contains the original data with one-hot encoded categorical features and no missing values

age  trestbps  chol  thalach  exang  oldpeak  target  sex_0  sex_1  cp_1  \
0     63       145   233      150      0      2.3       0    0.0    1.0   1.0   
1     67       160   286      108      1      1.5       1    0.0    1.0   0.0   
2     67       120   229      129      1      2.6       1    0.0    1.0   0.0   
3     37       130   250      187      0      3.5       0    0.0    1.0   0.0   
4     41       130   204      172      0      1.4       0    1.0    0.0   0.0   
..   ...       ...   ...      ...    ...      ...     ...    ...    ...   ...   
298   45       110   264      132      0      1.2       1    0.0    1.0   1.0   
299   68       144   193      141      0      3.4       1    0.0    1.0   0.0   
300   57       130   131      115      1      1.2       1    0.0    1.0   0.0   
301   57       130   236      174      0      0.0       1    1.0    0.0   0.0   
302   38       138   175      173      0      0.0       0    0.0    1.0   0.0   

     ...  slope_3  ca_0.0  ca_0.6722408026755853  ca_1.0  ca_2.0  ca_3.0  \
0    ...      1.0     1.0                    0.0     0.0     0.0     0.0   
1    ...      0.0     0.0                    0.0     0.0     0.0     1.0   
2    ...      0.0     0.0                    0.0     0.0     1.0     0.0   
3    ...      1.0     1.0                    0.0     0.0     0.0     0.0   
4    ...      0.0     1.0                    0.0     0.0     0.0     0.0   
..   ...      ...     ...                    ...     ...     ...     ...   
298  ...      0.0     1.0                    0.0     0.0     0.0     0.0   
299  ...      0.0     0.0                    0.0     0.0     1.0     0.0   
300  ...      0.0     0.0                    0.0     1.0     0.0     0.0   
301  ...      0.0     0.0                    0.0     1.0     0.0     0.0   
302  ...      0.0     0.0                    1.0     0.0     0.0     0.0   

     thal_3.0  thal_4.73421926910299  thal_6.0  thal_7.0  
0         0.0                    0.0       1.0       0.0  
1         1.0                    0.0       0.0       0.0  
2         0.0                    0.0       0.0       1.0  
3         1.0                    0.0       0.0       0.0  
4         1.0                    0.0       0.0       0.0  
..        ...                    ...       ...       ...  
298       0.0                    0.0       0.0       1.0  
299       0.0                    0.0       0.0       1.0  
300       0.0                    0.0       0.0       1.0  
301       1.0                    0.0       0.0       0.0  
302       1.0                    0.0       0.0       0.0  

[303 rows x 30 columns]

## Question No 4: Split the preprocessed data into training and testing sets (e.g., 80% training, 20% testing).

In [20]:
# Separate the features (X) and the target variable (y)
X = df_encoded.drop(columns=['target'])  # Features
y = heart_data_df['target']  # Target variable

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Part C Model Building 

## Question No 5: Choose and implement a classification / regression algorithm to build a heart disease prediction model. Train the model on the training data.

In [21]:
# Initialize the Random Forest Classifier
randomf_classifier = RandomForestClassifier(random_state=42)

# Train the model on the training data
randomf_classifier.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

## Question No 6: Evaluate the model's performance using appropriate classification / regression metrics (e.g., Mean Absolute Error, Root Mean Squared Error, confusion matrix, accuracy, recall, precision, f measure etc.). Provide a detailed explanation of the evaluation results.

In [22]:
# Analyzing these metrics to understanding the model's performance in predicting heart disease.
# Predict on the testing data
y_pred = randomf_classifier.predict(X_test)

In [23]:
# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.8688524590163934


In [24]:
# Calculate precision
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

Precision: 0.875


In [25]:
# Calculate recall
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

Recall: 0.875


In [26]:
# Calculate F1 score
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

F1 Score: 0.875


In [27]:
# Calculate confusion matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[25  4]
 [ 4 28]]


## Question No 7: Fine-tune the model by adjusting hyperparameters or trying different classification / regression algorithms.

### Using GridSearchCV

In [28]:
# Define the parameter grid to search
param_grid = {'n_estimators': [50, 100, 200],'max_depth': [None, 10, 20],'min_samples_split': [2, 5, 10]}

In [29]:
# Instantiate GridSearchCV
grid_search = GridSearchCV(estimator=randomf_classifier, param_grid=param_grid, cv=5)

In [30]:
# Fit the grid search to the data
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42),
             param_grid={'max_depth': [None, 10, 20],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [50, 100, 200]})

In [31]:
# Get the best parameters
best_params = grid_search.best_params_

In [32]:
# Train the model with the best hyperparameters
best_randomf_classifier = RandomForestClassifier(**best_params, random_state=42)
best_randomf_classifier.fit(X_train, y_train)

RandomForestClassifier(min_samples_split=10, n_estimators=200, random_state=42)

In [33]:
# Evaluate the model
y_pred_best = best_randomf_classifier.predict(X_test)
accuracy_best = accuracy_score(y_test, y_pred_best)

In [34]:
# print best hyperparameters and the Accuracy
print("Best Hyperparameters:", best_params)
print("Accuracy with Best Hyperparameters:", accuracy_best)

Best Hyperparameters: {'max_depth': None, 'min_samples_split': 10, 'n_estimators': 200}
Accuracy with Best Hyperparameters: 0.8852459016393442


### Using Random forest Classification Algorithm

In [35]:
model=RandomForestClassifier(n_estimators=100)
model.fit(X_train,y_train)
predict=model.predict(X_test)
print('Accuracy:', accuracy_score(y_test,predict))
print('ROC score',roc_auc_score(y_test,predict))
report = classification_report(y_test, predict)
print(report)
# Making report like this to be easy to understand the result

Accuracy: 0.8852459016393442
ROC score 0.8857758620689655
              precision    recall  f1-score   support

           0       0.87      0.90      0.88        29
           1       0.90      0.88      0.89        32

    accuracy                           0.89        61
   macro avg       0.88      0.89      0.89        61
weighted avg       0.89      0.89      0.89        61



### Using Gradient Boosting Classifier Algorithm

In [36]:
# Initialize and train Gradient Boosting Classifier
gb_classifier = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=3)
gb_classifier.fit(X_train, y_train)

GradientBoostingClassifier()

In [37]:
# Predictions on the test set
y_pred = gb_classifier.predict(X_test)

In [38]:
# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.8524590163934426


### Using cross-validation to estimate of the model's performance compared to another ways 

In [39]:
# Perform cross-validation
cv_scores = cross_val_score(gb_classifier, X, y, cv=5)  # 5-fold cross-validation

# Print the cross-validation scores
print("Cross-Validation Scores:", cv_scores)

Cross-Validation Scores: [0.85245902 0.90163934 0.80327869 0.78333333 0.76666667]


In [40]:
# Calculate the mean and standard deviation of cross-validation scores
mean_cv_score = np.mean(cv_scores)
std_cv_score = np.std(cv_scores)

print("Mean CV Score:", mean_cv_score)
print("Standard Deviation of CV Scores:", std_cv_score)

Mean CV Score: 0.8214754098360656
Standard Deviation of CV Scores: 0.04934810483135511


In [41]:
## finally we found the best accuracy by different algorithms and evaluation metrics (85%).

# Part D Reporting 

## a) Description of the Problem and Dataset: 
#The project aimed to build a machine learning model to predict heart disease. The dataset provided contains information about the patient attributes such as  age ,sex and colestrole rate , along with a binary target variable indicating whether the patient has a heart disease or not.

### Data contains;

1. age - age in years
2. sex - (1 = male; 0 = female)
3. cp - chest pain type
4. trestbps - resting blood pressure (in mm Hg on admission to the hospital)
5. chol - serum cholestoral in mg/dl
6. fbs - (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)
7. restecg - resting electrocardiographic results
8. thalach - maximum heart rate achieved
9. exang - exercise induced angina (1 = yes; 0 = no)
10. oldpeak - ST depression induced by exercise relative to rest
11. slope - the slope of the peak exercise ST segment
12. ca - number of major vessels (0-3) colored by flourosopy
13. thal - 3 = normal; 6 = fixed defect; 7 = reversable defect
14. target - have disease or not (1=yes, 0=no)

## b) Data preprocessing steps. 

1. Data Loading: The dataset was loaded using the fetch_ucirepo function, and its features and target were stored in a pandas DataFrame.
2. Handling Missing Values: Missing values were identified and imputed using the mean strategy for numerical features.
3. Encoding Categorical Features: Categorical features were encoded using one-hot encoding to convert them into a format suitable for machine learning algorithms.
4. Data Splitting: The preprocessed data was split into training and testing sets for model training and evaluation.


## c) Model selection, training, and evaluation. 
1. Model Selection: Two machine learning algorithms, namely Random Forest Classifier and Gradient Boosting Classifier, were selected for training due to their effectiveness in handling classification tasks and their ability to handle complex relationships in the data.
2. Training: The selected models were trained on the training data using default hyperparameters initially.
3. Evaluation: The models were evaluated using cross-validation to estimate their performance. Metrics such as accuracy, precision, recall, and F1-score were calculated to assess the models' performance. Additionally, a confusion matrix was generated to understand the distribution of true positives, true negatives, false positives, and false negatives.

### d) Discussion of Results and Insights Gained:

1. Results: The Random Forest Classifier achieved an accuracy of 85% on the test set, while Gradient Boosting Classifier achieved a slightly lower accuracy of 83%. Random Forest Classifier from the Random Forest Classification Algorithm achieved an accuracy of 84%.
2. Insights: The models performed reasonably well in predicting heart disease based on the given features. However, further analysis revealed that certain features might have a more significant impact on the prediction than others.

### e) Recommendations for Further Improvements:

1. Feature Engineering: Exploring additional features or engineering existing ones could potentially improve model performance.
2. Hyperparameter Tuning: Fine-tuning hyperparameters using advanced techniques like RandomizedSearchCV or Bayesian Optimization might lead to better model performance.
3. Model Ensemble: Ensemble methods such as bagging or boosting could be explored to combine the strengths of multiple models for improved prediction accuracy.
4. Handling Imbalanced Data: If the dataset is imbalanced, employing techniques like oversampling, undersampling, or using appropriate evaluation metrics for imbalanced classes could enhance model performance.
